pyclibrary/c_model.py

```diff
@@ -0,0 +1,779 @@
+# -*- coding: utf-8 -*-
+# -----------------------------------------------------------------------------
+# Copyright 2015 by PyCLibrary Authors, see AUTHORS for more details.
+#
+# Distributed under the terms of the MIT/X11 license.
+#
+# The full license is in the file LICENCE, distributed with this software.
+# -----------------------------------------------------------------------------
+"""Objects in this file can be used to model the public interface of a c
+library. The Rootclass is CLibInterface, which refers to all typedef-,
+macro-, variable-, function- objects contained in the library.
+
+Usually the user of the pyclibrary is not getting into contact with these
+objects. The CParser() is generating these objects automaticially from a
+c header file and the backend is relying on the definitions of these
+objects to now how to interact with the underlying c library.
+
+The class hierarchy is:
+
+* CLibInterface (a catalogue of all CLibBase objects in a library)
+* CLibBase (base class of all objects in header file)
+  * Macro
+    * ValMacro (#define x ...)
+    * FnMacro (#define x(...) ...)
+  * CLibType
+    * SimpleType
+      * BuiltinType (int, char, double, ...)
+      * CustomType (all references to custom types = typedefs or named
+                    AlgebraicDataTypes)
+    * AlgebraicDataType
+      * EnumType
+      * CompoundType
+        * StructType
+        * UnionType
+        * BitFieldType
+    * ComposedType
+      * PointerType
+      * ArrayType
+      * FunctionType
+"""
+from __future__ import (division, unicode_literals, print_function,
+                        absolute_import)
+import collections
+import itertools
+import re
+from future.moves.itertools import zip_longest
+from .errors import UnknownCustomTypeError
+
+
+def _lpadded_str(text):
+    """An internal helper, returns '' if text is None otherwise ' '+text"""
+    if text is None:
+        return ''
+    else:
+        return ' ' + text
+
+
+class CLibBase(object):
+
+    __slots__ = ()
+
+    def _getattrnames(self):
+        """Internal method that lists all fields
+        :rtype: list[str]
+        """
+        for cls in type(self).__mro__:
+            if hasattr(cls, '__slots__'):
+                for name in cls.__slots__:
+                    yield name
+
+    def copy(self):
+        """Creates a shallow copy o this type.
+        :rtype: CLibBase
+        """
+        attrs = {anm: getattr(self, anm) for anm in self._getattrnames()}
+        return type(self)(**attrs)
+
+    def __eq__(self, other):
+        if type(self) != type(other):
+            return False
+        return all(getattr(self, name) == getattr(other, name)
+                   for cls in type(self).__mro__
```

Comment:
Don't know how I could oversee this obvious code duplication (_getattrnames() was intended exactly for this purpose). Fixed it in 502716deca6cb58bedcd1658dffd8f73d17daa64.